### PR TITLE
Unit 4.1 edit profile refactor

### PR DESCRIPTION
**Pull request by nathanjophillips**
## The Purpose
- To learn more about how polymer components work and to grasp a better understanding of how data is being bound throughout the app.

## What It Does
1. This combines `edit-user` into `user-profile` then deletes `edit-user` to simplify some of the complexity with the file structure and to use less repetitive code.

2. The properties inside `user-profile` are currently not connected to a user object like they were before, but are on their own to be added to the `user-list` dom-repeat to the current item in the array.

3. The `user-profile` component now includes another template triggered by an editUserSelected() boolean function as well the same input functions that were used inside the original `edit-user` component.

## This accomplishes the following requirements given for  [Unit 4.1](https://github.com/phillnab/nathan-phillips-ux-iop/blob/master/issues/unit-4/4.1-polymer-components.md)
- The following HTML pages are represented in the Polymer app:
    - [x] user-list.html
    - [x] user-profile.html
    - [x] create-user.html
    - [x] edit-user.html
- [x] The `user-list` element has `user-profile` components nested inside it
- [x] The `user-profile` component maintains its expand and collapse functionality

# How To Test
- Clone Repo [nathan-phillips-ux-iop](https://github.com/phillnab/nathan-phillips-ux-iop)
- Checkout branch `nathan-phillips-ux-iop/unit-4.1-edit-profile-refactor`
- Run `yarn`
- Run `yarn build`
- Run `yarn start`

### Step 1 
- See if it matches this:

Screen Shot 2019-06-03 at 12 27 11 PM

### Step 2 
- Click `Collapse` to see if that functionally still works:

Screen Shot 2019-06-03 at 12 27 31 PM

### Step 3
- Click `edit` and change default data and `Save`:

Screen Shot 2019-06-03 at 12 27 42 PM

### Step 4 
- See if it Saved the input and added the new created user to the bottom of the userlist:

Screen Shot 2019-06-03 at 12 28 09 PM

## Checklist

Under penalty of public shaming, I avow that I:

- [x] Reviewed the diff to look for typos, whitespace errors, and console.log() statements
- [x] Verified that there are no compiler or linter errors or warnings
- [ ] Verified the build in production(optimized) mode
- [x] Updated the docs, if necessary
- [x] Included the appropriate labels
- [x] Referenced applicable issues (i.e. Closes #XXXX, Fixes #XXXX)

Browsers tested:
- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge
- [ ] Internet Explorer

### CODE REVIEW

File: src/components/create-user.html [hunk 1]
```diff
@@ -44,31 +42,90 @@
...
+            value: false
+          },
+          user: {
+            type: Object,
+            value: function() {
+              return {
+                id: "0",
+                firstName: '',
+                lastName: '',
+                phoneNumber: 0,
```
Comment (Jropp) [inline src/components/create-user.html]: This displays a zero when you go to create a new user

File: src/components/create-user.html [hunk 2]
```diff
@@ -44,31 +42,90 @@
...
+          <td><label for="firstName">First name :</label></td>
+          <td><input type="text" name="firstName" id="firstName" value="[[user.firstName]]" on-input="updateFirstName"></td>
+        </tr>
+        <tr>
+          <td><label for="lastName">Last name :</label></td>
+          <td><input type="text" name="lastName" id="lastName" value="[[user.lastName]]" on-input="updateLastName"></td>
+        </tr>
+        <tr>
+          <td><label for="phone">Phone :</label></td>
+          <td><input type="text" name="phone" id="phone" value="[[user.phoneNumber]]" on-input="updatePhoneNumber"></td>
```
Comment (Jropp) [inline src/components/create-user.html:60]: You have this as type="text" but in the property you are setting it as a number

File: src/components/intern-app.html [hunk 3]
```diff
@@ -44,6 +44,41 @@ <h1>Intern App</h1>
...
+        window.addEventListener("newUser", this.boundNewUser);   
+      }
+
+      disconnectedCallback() {
+        super.disconnectedCallback();
+        window.removeEventListener("newUser", this.boundNewUser); 
+      }
+
+      newUser(event) {
+        var newUserProfile = Object.assign({},event.detail);
```
Comment (Jropp) [inline src/components/intern-app.html]: Space after the comma

File: src/components/intern-app.html [hunk 4]
```diff
@@ -44,6 +44,41 @@ <h1>Intern App</h1>
...
+        newUserProfile.id = this.genId();
+        this.pushToArray(this.userProfiles, newUserProfile);
+      }
+
+      pushToArray(userProfiles, newUserProfile) {
+        var existingIds = userProfiles.map((newUserProfile) => newUserProfile.id);
+        if (!existingIds.includes(newUserProfile.id)) {
```
Comment (Jropp) [inline src/components/intern-app.html:70]: I think you could skip the need for a .map and a .includes if you used .find with a function as your argument

```js
userProfiles.find(profile => {
  // some stuff in here
}) 
```



https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find

File: src/components/create-user.html (hunk 1 above)
Comment (nathanjophillips) [reply src/components/create-user.html]: switched it to a string

File: src/components/intern-app.html [hunk 5]
```diff
@@ -15,8 +15,8 @@
     <header>
       <h1>Intern App</h1>
     </header>
-    <edit-user user="[[userProfiles.0]]"></edit-user>
-    <user-list user-profiles="[[userProfiles]]"></user-list>
+    <create-user></create-user>
+    <user-list user="[[userProfiles.0]]" user-profiles="[[userProfiles]]"></user-list>
```
Comment (collincahill) [inline src/components/intern-app.html]: I don't think `user="[[userProfiles.0]]"` is necessary anymore 😄

**Review by collincahill — COMMENTED** [review]
Just a note: the entire app is left-aligned (not centered) in Firefox

**Comment by Jropp** [comment]
I'm also getting the firefox issue, which also distorts the edit user mode a bit.

Screen Shot 2019-06-04 at 9 00 40 PM

File: src/components/intern-app.html [hunk 6]
```diff
@@ -15,8 +15,8 @@
     <header>
       <h1>Intern App</h1>
     </header>
-    <edit-user user="[[userProfiles.0]]"></edit-user>
-    <user-list user-profiles="[[userProfiles]]"></user-list>
+    <create-user></create-user>
+    <user-list  user-profiles="[[userProfiles]]"></user-list>
```
Comment (Jropp) [inline src/components/intern-app.html]: White space needs deleted

**Review by Jropp — APPROVED** [review]
Du du du done

**Pull request merged**